src/main/scala/workflow/graph/GraphExecutor.scala

```diff
@@ -0,0 +1,110 @@
+package workflow.graph
+
+/**
+ * A GraphExecutor is constructed using a graph with already-executed state attached to it.
+ * It provides methods to further execute parts of the graph, returning the execution result.
+ * By default, it will optimize the graph before any new execution occurs.
+ *
+ * Warning: Not thread-safe!
+ *
+ * @param graph The underlying graph to execute
+ * @param state A mapping of GraphId to computed Expression.
+ *              This is the current state of what has already been executed in the graph.
+ * @param optimize Whether or not to optimize the graph using the global executor before executing anything.
+ *                 Defaults to True.
+ */
+private[graph] class GraphExecutor(graph: Graph, state: Map[GraphId, Expression], optimize: Boolean = true) {
```

Comment:
GraphExecutor: Given a graph & execution state, provides methods that will optimize the graph & execute it up to arbitrary GraphIds.

GraphExecution: Lazily represent the result of executing a pipeline on data. Given a GraphExecutor, a desired sink, and optionally sources that still need to be inserted into the graph, do the two-step process I describe above to get the result of executing the pipeline on data.